packages/discord.js-utilities/src/lib/builders/MessageBuilder.ts

```diff
@@ -0,0 +1,168 @@
+import type { MessageEmbed, MessageMentionOptions, MessageOptions } from 'discord.js';
+
+/**
+ * Exclude null and undefined from T
+ */
+type Defined<T> = T extends undefined ? never : T;
+
+export type MessageBuilderFileResolvable = Defined<MessageOptions['files']>[number];
+export type MessageBuilderCodeResolvable = Defined<MessageOptions['code']>;
+export type MessageBuilderSplitResolvable = Defined<MessageOptions['split']>;
+export type MessageBuilderResolvable = Omit<MessageOptions, 'embed' | 'disableMentions' | 'reply'> & { embed?: MessageEmbed };
+
+/**
+ * A message builder class, it implements the {@link https://discord.js.org/#/docs/main/stable/typedef/MessageOptions MessageOptions}
+ * interface.
+ */
+export class MessageBuilder implements MessageOptions {
+	/**
+	 * Whether or not the message should be spoken aloud.
+	 * @default false
+	 */
+	public tts?: boolean;
+
+	/**
+	 * The nonce for the message.
+	 * @default ''
+	 */
+	public nonce?: string;
+
+	/**
+	 * The content for the message. If set to undefined and the builder is used to edit, the content will not be
+	 * replaced.
+	 */
+	public content?: string;
+
+	/**
+	 * An embed for the message. If set to undefined and the builder is used to edit, the embed will not be replaced.
+	 */
+	public embed?: MessageEmbed;
+
+	/**
+	 * Which mentions should be parsed from the message content.
+	 */
+	public allowedMentions?: MessageMentionOptions;
+
+	/**
+	 * Files to send with the message. This should not be set when editing a message, as Discord does not support
+	 * editing file attachments.
+	 */
+	public files?: MessageBuilderFileResolvable[];
+
+	/**
+	 * Language for optional codeblock formatting to apply.
+	 */
+	public code?: MessageBuilderCodeResolvable;
+
+	/**
+	 * Whether or not the message should be split into multiple messages if it exceeds the character limit. If an object
+	 * is provided, these are the options for splitting the message.
+	 */
+	public split?: MessageBuilderSplitResolvable;
+
+	public constructor(options?: MessageBuilderResolvable) {
+		this.tts = options?.tts;
+		this.nonce = options?.nonce;
+		this.content = options?.content;
+		this.embed = options?.embed;
+		this.allowedMentions = options?.allowedMentions;
+		this.files = options?.files;
+		this.code = options?.code;
+		this.split = options?.split;
```

Comment:
theres no defaults set
also add a default options static object we can use (think always setting allowedmentions to none for example)

Reply:
And... what do you propose the default to be? We can't make the default be a MessageBuilder because we need to construct the class itself which would require the static field to be set and constructed...

Reply:
Make them messageOptions too! Just set everything to null, or to reasonable defaults (no allowed mentions, no tts, you'll figure it out)

Reply:
How 'bout everything `undefined`, so the defaults are Discord's defaults?

Reply:
Deal

Reply:
Fixed in tts
nonce
content
embed
allowedMentions
files
code
split